src/py/flwr/cli/log.py

```diff
@@ -54,16 +55,45 @@ def start_stream(
         channel.close()
 
 
-def stream_logs(run_id: int, channel: grpc.Channel, duration: int) -> None:
-    """Stream logs from the beginning of a run with connection refresh."""
-    start_time = time.time()
-    stub = ExecStub(channel)
-    req = StreamLogsRequest(run_id=run_id)
+def stream_logs(
+    run_id: int, stub: ExecStub, duration: int, after_timestamp: float
+) -> float:
+    """Stream logs from the beginning of a run with connection refresh.
+
+    Parameters
+    ----------
+    run_id : int
+        The identifier of the run.
+    stub : ExecStub
+        The gRPC stub to interact with the Exec service.
+    duration : int
+        The timeout duration for each stream connection in seconds.
+    after_timestamp : float
+        The timestamp to start streaming logs from.
+
+    Returns
+    -------
+    float
+        The latest timestamp from the streamed logs or the provided `after_timestamp`
+        if no logs are returned.
+    """
+    req = StreamLogsRequest(run_id=run_id, after_timestamp=after_timestamp)
+
+    latest_timestamp = 0.0
+    res = None
+    try:
+        for res in stub.StreamLogs(req, timeout=duration):
+            print(res.log_output, end="")
+    except grpc.RpcError as e:
+        # pylint: disable=E1101
+        if e.code() != grpc.StatusCode.DEADLINE_EXCEEDED:
+            raise e
+    finally:
+        if res is not None:
+            latest_timestamp = cast(float, res.latest_timestamp)
 
-    for res in stub.StreamLogs(req):
-        print(res.log_output)
-        if time.time() - start_time > duration:
-            break
+    # Add a small epsilon to the latest timestamp to avoid getting the same log
+    return max(latest_timestamp + 1e-6, after_timestamp)
```

Comment:
Shouldn't the SuperLink be able to figure this out deterministically? We send the timestamp of the last log we have and the server returns the next log with a timestamp greater than the timestamp we sent.

Reply:
I'd like to allow users to fetch all logs after a certain timestamp, and thus I added a `after_timestamp` instead of `latest_log_timestamp` to the `StreamLogsRequest`.

Reply:
Given that, it's better to set `after_timestamp` to `latest_log_timestamp + 1e-6` to stream following logs.